### PR TITLE
[clang] Add tests for CWG1800-1804

### PR DESCRIPTION
**Pull request by Endilll**
Covers C++ core issues 1800, 1801, 1802, 1803, 1804.

### CODE REVIEW

**Comment by Endilll** [comment]
This PR also updates status for CWG437, which is a follow-up to #77372.
Nothing wrong with the tooling, just my mistake separating output of make_cxx_dr_status from changes I already had for complete-class context DRs.

File: clang/test/CXX/drs/dr18xx.cpp [hunk 1]
```diff
@@ -1,16 +1,169 @@
...
+// cxx98-14-error@-1 {{non-type template argument does not refer to any declaration}}
+//   cxx98-14-note@#dr1801-S {{template parameter is declared here}}
+// since-cxx17-error@#dr1801-S-i {{non-type template argument refers to subobject '.i'}}
+}
+
+namespace dr1802 { // dr1802: 3.1
+#if __cplusplus >= 201103L
+// Using a Wikipedia example of surrogate pair:
+// https://en.wikipedia.org/wiki/UTF-16#Examples
+constexpr char16_t a[3] = u"𐐷";
```
Comment (cor3ntin) [inline clang/test/CXX/drs/dr18xx.cpp:35]: This is really just an array of `char16_t` - so `𐐷` would be parsed as 2 char16_t
What the wording is referring to is that any of the following should yield 2 elements 

```cpp
u"\N{DESERET SMALL LETTER YEE}"
u"\u{10437}"
u"\U00010437"
```

Comment (cor3ntin) [reply clang/test/CXX/drs/dr18xx.cpp:35]: https://eel.is/c++draft/lex.charset#nt:universal-character-name

File: clang/test/CXX/drs/dr18xx.cpp [hunk 2]
```diff
@@ -1,16 +1,169 @@
...
+static_assert(a[0] == 0xD801, "");
+static_assert(a[1] == 0xDC37, "");
+static_assert(a[2] == 0x0, "");
+#endif
+} // namespace dr1802
+
+namespace dr1803 { // dr1803: 2.9
+#if __cplusplus >= 201103L
```
Comment (cor3ntin) [inline clang/test/CXX/drs/dr18xx.cpp:43]: @erichkeane I think it looks reasonable but I would like a second pair of eyes
https://cplusplus.github.io/CWG/issues/1804.html

File: clang/test/CXX/drs/dr18xx.cpp (hunk 1 above)
Comment (Endilll) [reply clang/test/CXX/drs/dr18xx.cpp]: Thank you! Updated the test to use `\U00010437` as the first two forms you provided are too new for older language modes.

File: clang/test/CXX/drs/dr18xx.cpp (hunk 2 above)
Comment (Endilll) [reply clang/test/CXX/drs/dr18xx.cpp:43]: (Test for 1804 is at line 56)

Comment (erichkeane) [reply clang/test/CXX/drs/dr18xx.cpp:43]: Yeah, I think the test for 1804 tests everything I can thing of.

**Review by erichkeane — APPROVED** [review]

**Review by cor3ntin — APPROVED** [review]
LGTM

**Pull request merged**